### ISSUE TITLE
[BUG] Finding installed modules failed SyntaxError: Unexpected end of JSON input

### ISSUE THREAD

**Issue by piotrwitek** [Microsoft/vscode-npm-scripts#27]
I got error whenever I open my package.json in the editor:
Finding installed modules failed SyntaxError: Unexpected end of JSON input

![screen shot 2016-12-18 at 16 10 45](https://cloud.githubusercontent.com/assets/739075/21294463/41fb60ea-c53d-11e6-8753-0b89d8d7474f.png)

When disabled your extension error disappeared.

**Comment by piotrwitek** [comment]
My Scripts: 
```"scripts": {
    "precommit": "yarn run lint & yarn run tsc",
    "prepush": "yarn run test",
    "rebuild": "yarn run clean && yarn && yarn run build",
    "clean": "rm -rf node_modules dist",
    "dev": "webpack-dev-server --config configs/webpack.config.js",
    "build": "NODE_ENV=production webpack -p --config configs/prod.config.js --colors",
    "prod": "webpack-dev-server --config configs/webpack.config.js",
    "lint": "tslint 'client/**/*.ts' 'client/**/*.tsx'",
    "tsc": "tsc -p client/ --noEmit",
    "test": "jest --config ./jest.config.json",
    "test:update": "jest -u  --config ./jest.config.json",
    "docker:build": "docker build -t mad_app_image .",
    "docker:run": "docker run --rm --name mad-app -p 3000:3000 mad_app_image",
    "docker": "npm run docker:build && npm run docker:run"
  },
```

**Comment by egamma** [comment]
@piotrwitek thanks for the issue report. Published a new version with the fix to the marketplace.

**Comment by piotrwitek** [comment]
@egamma Thanks for fast response but the issue is not fixed, updated to 0.1.3 and still exactly the same error

**Comment by egamma** [comment]
@piotrwitek can you please attach the full package.json, so that I can reproduce your exception please.

**Comment by egamma** [comment]
@piotrwitek one more thought do you have npm installed or do you only use yarn? The validation calls `npm ls` which assumes that npm is installed.

**Comment by egamma** [comment]
Could reproduce the issue when 'npm' is not installed. The code no handles this case gracefully. Published version 0.1.4 to the marketplace.

**Comment by piotrwitek** [comment]
@egamma Thanks, seems like 0.1.4 fixed exception.
 Regarding your questions:
- I have `npm` and `npm ls` is working fine.
Your fix seems to just handle the exception, isn't that `getInstalledModules` logic is not handling something that it have raised exception in the first place?

**Comment by egamma** [comment]
@piotrwitek yes, the fix is brute force and swallows all exceptions that happen when running `npm ls --depth 0 --json`. 

This includes 1) spawning the process fails or 2) the returned JSON cannot be parsed. I could reproduce 1) but I failed to reproduce 2). For that I'd need your package.json  or the output your get when you run  `npm ls --depth 0 --json` in the folder with the package.json that is causing trouble.

**Comment by piotrwitek** [comment]
@egamma Cool, Here you go: 
```
{
  "name": "mad-app",
  "version": "0.1.0",
  "private": true,
  "description": "Description",
  "author": "Piotr Witek <piotr.witek@shipwallet.com> (http://piotrwitek.github.io)",
  "license": "SEE LICENSE IN LICENSE.md",
  "engines": {
    "node": ">=6.9.0"
  },
  "scripts": {
    "precommit": "yarn run lint & yarn run tsc",
    "prepush": "yarn run test",
    "clean": "rm -rf node_modules dist",
    "build": "NODE_ENV=production webpack --config configs/prod.config.js",
    "rebuild": "yarn run clean && yarn && yarn run build",
    "prod": "webpack-dev-server --config configs/webpack.config.js",
    "dev": "webpack-dev-server --config configs/webpack.config.js",
    "lint": "tslint 'client/**/*.ts' 'client/**/*.tsx'",
    "tsc": "tsc -p client/ --noEmit",
    "test": "jest --config ./jest.config.json",
    "test:update": "jest -u  --config ./jest.config.json",
    "docker:build": "docker build -t mad_app_image .",
    "docker:run": "docker run --rm --name mad-app -p 3000:3000 mad_app_image",
    "docker": "npm run docker:build && npm run docker:run"
  },
  "dependencies": {
    "@blueprintjs/core": "^1.4.0",
    "antd": "^2.5.2",
    "blaze": "^3.2.0",
    "classnames": "2.2.5",
    "jwt-decode": "2.1.0",
    "react": "^15.4.1",
    "react-dom": "^15.4.1",
    "react-icons": "2.2.1",
    "react-redux": "^4.4.6",
    "react-router": "3.0.0",
    "react-router-redux": "4.0.7",
    "react-select": "^1.0.0-rc.2",
    "redux": "^3.6.0",
    "redux-actions": "^1.2.0",
    "redux-thunk": "^2.1.0",
    "reselect": "2.5.4",
    "seamless-immutable": "6.3.0",
    "uuid": "^3.0.1",
    "whatwg-fetch": "2.0.1"
  },
  "devDependencies": {
    "@types/dom4": "1.5.20",
    "@types/enzyme": "^2.5.39",
    "@types/expect": "^1.13.31",
    "@types/jest": "^16.0.1",
    "@types/jwt-decode": "1.4.28",
    "@types/node": "^6.0.52",
    "@types/react": "^0.14.51",
    "@types/react-dom": "0.14.19",
    "@types/react-hot-loader": "^3.0.0",
    "@types/react-redux": "^4.4.34",
    "@types/react-router": "^2.0.40",
    "@types/react-router-redux": "^4.0.35",
    "@types/react-select": "^1.0.36",
    "@types/redux-actions": "^0.8.34",
    "@types/redux-thunk": "2.1.32",
    "@types/seamless-immutable": "6.1.2",
    "@types/whatwg-fetch": "^0.0.33",
    "compression": "1.6.2",
    "connect-history-api-fallback": "1.3.0",
    "copy-webpack-plugin": "^4.0.1",
    "css-loader": "^0.26.1",
    "css-modules-require-hook": "4.0.5",
    "enzyme": "2.6.0",
    "expect": "^1.20.2",
    "express": "4.14.0",
    "extract-text-webpack-plugin": "^2.0.0-beta.4",
    "file-loader": "0.9.0",
    "html-webpack-plugin": "^2.24.1",
    "http-proxy": "^1.16.2",
    "husky": "0.11.9",
    "image-webpack-loader": "^3.1.0",
    "jest": "^17.0.3",
    "json-loader": "0.5.4",
    "less": "^2.7.1",
    "less-loader": "^2.2.3",
    "node-sass": "^3.13.0",
    "react-addons-css-transition-group": "^15.4.1",
    "react-addons-test-utils": "^15.4.1",
    "react-hot-loader": "3.0.0-beta.6",
    "react-test-renderer": "^15.4.1",
    "redbox-react": "^1.3.3",
    "redux-logger": "^2.7.4",
    "sass-loader": "4.0.2",
    "source-map-loader": "0.1.5",
    "style-loader": "0.13.1",
    "ts-jest": "^17.0.3",
    "ts-loader": "^1.3.3",
    "tslint": "^4.1.1",
    "typescript": "^2.1.4",
    "url-loader": "0.5.7",
    "webpack": "^2.2.0-rc.1",
    "webpack-dev-server": "^2.2.0-rc.0"
  }
}
```
[npm-ls.txt](https://github.com/Microsoft/vscode-npm-scripts/files/659725/npm-ls.txt)

**Comment by egamma** [comment]
@piotrwitek thanks. 

In my setup (without yarn installed) I cannot reproduce the error.  The screenshot below illustrates the result of a validation (I've manually uninstalled the `antd` module to force a validation warning).

The exception happens when the output from `npm ls --depth 0 --json` cannot be parsed as JSON.  I noticed that `yarn` doesn't provide the `--json` command line argument. Speculation, if `yarn` sets up an alias for npm then this could happen. I'll try to reproduce with `yarn` installed.

![validating](https://cloud.githubusercontent.com/assets/172399/21305253/dd83f094-c5ca-11e6-8811-1e6544cdc2d3.png)

**Comment by piotrwitek** [comment]
@egamma I've seen your lib uses npm bin name to run commands, and when I use `which npm` I only got npm path, so it makes it impossible to run yarn in such set up, we can repel such scenario.

Other thing is, if you look at npm-ls.txt dump I have attached above, there you will see some error in the returned json somewhere at the end.
The error is due to UNMET PEER DEP for  webpack2.2.0-rc.1, while in reality it shouldn't happened because it don't seem to have any peers set-up in package json.
I don't know if you already noticed that, but it would be the first place I would investigate.

**Comment by egamma** [comment]
@piotrwitek 

I've modified the code so that it `requires` your`npm-ls.txt` as JSON instead of running `npm -ls`, and the code could handle the JSON without any issues. This indicates that the problem isn't an issue of the JSON, but the way the output from the spawned process is read.

I've changed the code to do JSON parsing on the [close event](https://nodejs.org/api/child_process.html#child_process_event_close) instead of the [exit event](https://nodejs.org/api/child_process.html#child_process_event_exit). According the docs, it is possible that stdout still is open when `exit` is called. This is the [change](https://github.com/Microsoft/vscode-npm-scripts/blob/master/src/main.ts#L568).

I've attached a new version as a VSIX and would appreciate if you can try out whether this addresses the issue.
[vscode-npm-script-0.1.4.zip](https://github.com/Microsoft/vscode-npm-scripts/files/661592/vscode-npm-script-0.1.4.zip)

To try it:
- extract the file
- rename the extension from `zip` to `vsix`
- install the extension from the extension viewlets menu's 'Install VSIX' command (see below)

![image](https://cloud.githubusercontent.com/assets/172399/21320508/97d6d164-c611-11e6-90d5-a71a45335a08.png)
 
Regarding the 'unmet peer dep' the validation currently doesn't handle missing peer dependencies so this errors are ignored. Adding warnings for unmet peer dependencies would make sense.

**Comment by piotrwitek** [comment]
@egamma I done the test.
When using custom extension I got this error:
`[npm-script-runner] Cannot validate the package.json TypeError: Cannot read property 'version' of undefined`

Even when I empty my entire package.json it's still showing this strange error.

**Comment by egamma** [comment]
@piotrwitek thanks for trying and for your patience.

I've verified that the attached vsix works fine in my setup. I've attached a new version of the vsix that dumps the stack trace to the console when this exception happens. Can you please try again?

[vscode-npm-script-0.1.4.zip](https://github.com/Microsoft/vscode-npm-scripts/files/663597/vscode-npm-script-0.1.4.zip)

Here are the steps:
- manually uninstall the existing npm-script-runner
- rename the attached .zip to .vsix and install it as above
- when the exception occurs then open the Developer Tools: Help>Toggle Developer Tools and switch to the Console tab. The stack trace of the exception should show up there (see below)

![image](https://cloud.githubusercontent.com/assets/172399/21349590/00b0d1ca-c6b3-11e6-9f47-b8d49458078a.png)

**Comment by egamma** [comment]
>Even when I empty my entire package.json it's still showing this strange error.

This was caused by the assumption that a package.json has a least a `name` attribute.  This is fixed but it looks like a separate issue than the one you reported above:
> `[npm-script-runner] Cannot validate the package.json TypeError: Cannot read property 'version' of undefined`

**Comment by piotrwitek** [comment]
@egamma cool, I gonna test it later today so stay tuned :)

**Comment by egamma** [comment]
@piotrwitek it is not yet publishd to the marketplace, but here is the vsix with the fixes. Thanks in advance for checking. If there is the exception pls open the Dev tools console.

[vscode-npm-script-0.1.4.zip](https://github.com/Microsoft/vscode-npm-scripts/files/666135/vscode-npm-script-0.1.4.zip)

To install same steps as above https://github.com/Microsoft/vscode-npm-scripts/issues/27#issuecomment-268224592

**Comment by piotrwitek** [comment]
No problem, here is the stacktrace:
```
TypeError: Cannot read property 'version' of undefined
    at forEach.each (/Users/piotrek/.vscode/extensions/eg2.vscode-npm-script-0.1.4/out/src/main.js:283:61)
    at Array.forEach (native)
    at getDiagnostic (/Users/piotrek/.vscode/extensions/eg2.vscode-npm-script-0.1.4/out/src/main.js:275:41)
    at /Users/piotrek/.vscode/extensions/eg2.vscode-npm-script-0.1.4/out/src/main.js:234:38
    at next (<anonymous>)
    at fulfilled (/Users/piotrek/.vscode/extensions/eg2.vscode-npm-script-0.1.4/out/src/main.js:4:58)
```

**Comment by egamma** [comment]
@piotrwitek thanks this is helpful. 

This means that `npm ls` has found a dependency with an invalid version, but when I use a JSON parser  to locate the source position of a dependency in the JSON file it cannot find it. I know how to protect against this case, but I'd me more interested in finding the root cause of this failure. 

I've tried to create such a situation but I failed up to now. It seems that it is easy for you ;-). Therefore can please attach the package.json and the output from `npm ls --depth 0 --json` when this happens? I've analysed the output from npm-ls that you attached in this comment https://github.com/Microsoft/vscode-npm-scripts/issues/27#issuecomment-267852412 but this output did not have an 'invalid' installed version of a dependency.

**Comment by piotrwitek** [comment]
@egamma yes I can help you with that.
I'm not sure how to hook up a debugger to the extension in running vs code, so I could set a breakpoint, is there a way to do that? Could you point me to some article with example how to do it?

**Comment by egamma** [comment]
@piotrwitek thanks for your offer, but before we do the extension debugging step (which is well supported), let's do one more thing. Now, that we know where the problem is, I've instrumented the code to print out for which module the source range cannot be located. The module name of the problematic dependency is printed to the console. Therefore can you please:
- install this version (0.1.5) of the extension attached below
- open the package.json that triggered the exception before
- open the console in the dev tools and you should see the print out of the module name
- can you please include this module name, and the corresponding excerpt of the package.json in a comment.
- also include the snippet from `npm ls --depth 0 --json` with the `invalid` warning for this module, e.g.
```json
    "jwt-decode": {
      "version": "2.1.0",
      "from": "jwt-decode@2.1.0",
      "resolved": "https://registry.npmjs.org/jwt-decode/-/jwt-decode-2.1.0.tgz",
      "invalid": true,
      "problems": [
        "invalid: jwt-decode@2.1.0 C:\\Users\\egamma\\Projects\\tries\\npm-script-test\\node_modules\\jwt-decode"
      ]
    },
```

[vscode-npm-script-0.1.5.zip](https://github.com/Microsoft/vscode-npm-scripts/files/670541/vscode-npm-script-0.1.5.zip)

**Comment by piotrwitek** [comment]
Done, the problem was:
- [npm-script] Could not locate "invalid" dependency 'rc-select' in package.json
- I found rc-select package folder in node_modules root to be empty (so no package.json was found inside)
- rc-select was also located in antd node_modules, but correctly installed there with all contents
- after removing node_modules and reinstalling with npm install - fixed the issue

I believe the problem was that if package.json of module is missing then this extension was throwing a null exception due to deep object property access, without properly checking if they exist.
Adding error handling for deep property access should fix the future issues. 

Now extension was working fine without exception, so I believe you have already fixed that in 0.1.5

**Comment by egamma** [comment]
@piotrwitek thanks for going through this with me and helping me to make the code more robust.

I was assuming that an `invalid` entry in an `npm ls` JSON report always has a `version` property, but in the case of an empty package directory, the version cannot be determined and npm correctly reports an error and there is no `version` property. 

I've published the new version `0.1.5` to the market place. To get future updates of the module, you need to uninstall the manually installed extension and install it from the extension viewlet again.

**Comment by piotrwitek** [comment]
@egamma no problem that was cool, I learned something new about VS Code! :) BR
